pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -787,19 +826,26 @@ func (nc *Controller) monitorNodeHealth() error {
 				nc.processNoTaintBaseEviction(node, &observedReadyCondition, gracePeriod)
 			}
 
-			// Report node event.
-			if currentReadyCondition.Status != v1.ConditionTrue && observedReadyCondition.Status == v1.ConditionTrue {
+			// We call MarkPodsNotReady when node condition changed from true ti something different
+			// or provious MarkPodsNotReady and ready status is not true.
+			if (currentReadyCondition.Status != v1.ConditionTrue && observedReadyCondition.Status == v1.ConditionTrue) ||
+				(nc.retryMarkPodsNotReadySet.has(node.Name) && observedReadyCondition.Status != v1.ConditionTrue) {
+				// Report node event.
 				nodeutil.RecordNodeStatusChange(nc.recorder, node, "NodeNotReady")
 				pods, err := listPodsFromNode(nc.kubeClient, node.Name)
 				if err != nil {
 					utilruntime.HandleError(fmt.Errorf("Unable to list pods from node %v: %v", node.Name, err))
+					nc.retryMarkPodsNotReadySet.insert(node.Name)
 					continue
 				}
 				if err = nodeutil.MarkPodsNotReady(nc.kubeClient, pods, node.Name); err != nil {
 					utilruntime.HandleError(fmt.Errorf("Unable to mark all pods NotReady on node %v: %v", node.Name, err))
+					nc.retryMarkPodsNotReadySet.insert(node.Name)
+					continue
 				}
 			}
 		}
+		nc.retryMarkPodsNotReadySet.delete(node.Name)
```

Comment:
This stuff requires a unit test so that we don't regress in the future.

Reply:
Done. `TestMonitorNodeHealthMarkPodsNotReadyRetry`